File: src/tools/troubleshoot.ts
import { exec } from "child_process";
import { promisify } from "util";

const execAsync = promisify(exec);

interface TroubleshootDeploymentArgs {
  namespace: string;
  releaseName?: string;
  checkType?: "all" | "pods" | "logs" | "helm" | "route" | "certificates" | "vault-status";
  podName?: string;
}

export async function troubleshootDeployment(args: any) {
  const {
    namespace,
    releaseName,
    checkType = "all",
    podName,
  } = args as TroubleshootDeploymentArgs;

  try {
    let response = `# Troubleshooting: ${namespace}\n\n`;
    
    if (checkType === "all" || checkType === "pods") {
      response += await checkPods(namespace, releaseName);
    }
    
    if (checkType === "all" || checkType === "logs") {
      if (podName) {
        response += await getPodLogs(namespace, podName);
      } else if (releaseName) {
        response += await getPodLogs(namespace, `${releaseName}-0`);
      }
    }
    
    if (checkType === "all" || checkType === "helm") {
      if (releaseName) {
        response += await checkHelmRelease(namespace, releaseName);
      }
    }
    
    if (checkType === "all" || checkType === "route") {
      if (releaseName) {
        response += await checkRoute(namespace, releaseName);
      }
    }
    
    if (checkType === "vault-status") {
      if (releaseName) {
        response += await checkVaultStatus(namespace, releaseName);
      }
    }

    response += `\n## Remediation Suggestions\n\n`;
    response += await generateRemediationSuggestions(namespace, releaseName);

    return {
      content: [{ type: "text", text: response }],
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      content: [
        {
          type: "text",
          text: `Troubleshooting failed: ${errorMessage}`,
        },
      ],
      isError: true,
    };
  }
}

async function checkPods(namespace: string, releaseName?: string) {
  let command = `oc get pods -n ${namespace}`;
  if (releaseName) {
    command += ` -l app.kubernetes.io/instance=${releaseName}`;
  }
  command += ` -o wide`;

  try {
    const { stdout } = await execAsync(command);
    return `## Pod Status\n\n\`\`\`\n${stdout}\`\`\`\n\n`;
  } catch (error) {
    return `## Pod Status\n\n❌ Unable to retrieve pod status\n\n`;
  }
}

async function getPodLogs(namespace: string, podName: string) {
  const command = `oc logs -n ${namespace} ${podName} --tail=100`;

  try {
    const { stdout } = await execAsync(command);
    return `## Pod Logs: ${podName}\n\n\`\`\`\n${stdout}\`\`\`\n\n`;
  } catch (error) {
    return `## Pod Logs: ${podName}\n\n❌ Unable to retrieve logs\n\n`;
  }
}

async function checkHelmRelease(namespace: string, releaseName: string) {
  const command = `helm status ${releaseName} -n ${namespace}`;

  try {
    const { stdout } = await execAsync(command);
    return `## Helm Release Status\n\n\`\`\`\n${stdout}\`\`\`\n\n`;
  } catch (error) {
    return `## Helm Release Status\n\n❌ Release not found or error occurred\n\n`;
  }
}

async function checkRoute(namespace: string, releaseName: string) {
  const command = `oc get route -n ${namespace} ${releaseName}-route -o wide`;

  try {
    const { stdout } = await execAsync(command);
    return `## Route Status\n\n\`\`\`\n${stdout}\`\`\`\n\n`;
  } catch (error) {
    return `## Route Status\n\n❌ Route not found\n\n`;
  }
}

async function checkVaultStatus(namespace: string, releaseName: string) {
  const podName = `${releaseName}-0`;
  const command = `oc exec -n ${namespace} ${podName} -- vault status`;

  try {
    const { stdout } = await execAsync(command);
    return `## Vault Status\n\n\`\`\`\n${stdout}\`\`\`\n\n`;
  } catch (error) {
    return `## Vault Status\n\n❌ Unable to check vault status (pod may not be ready)\n\n`;
  }
}

async function generateRemediationSuggestions(namespace: string, releaseName?: string) {
  let suggestions = ``;
  
  // Check common issues
  try {
    const { stdout } = await execAsync(`oc get pods -n ${namespace} -o json`);
    const pods = JSON.parse(stdout);
    
    if (pods.items && pods.items.length === 0) {
      suggestions += `- No pods found in namespace. Check if deployment was successful.\n`;
    } else {
      pods.items.forEach((pod: any) => {
        if (pod.status.phase !== "Running") {
          suggestions += `- Pod ${pod.metadata.name} is in ${pod.status.phase} state\n`;
          
          if (pod.status.containerStatuses) {
            pod.status.containerStatuses.forEach((container: any) => {
              if (container.state.waiting) {
                suggestions += `  - Container waiting: ${container.state.waiting.reason}\n`;
              }
            });
          }
        }
      });
    }
  } catch (error) {
    suggestions += `- Unable to analyze pod status automatically\n`;
  }

  if (!suggestions) {
    suggestions = `- All checks passed. If issues persist, check application logs.\n`;
  }

  return suggestions;
}

File: python_src/tools/troubleshoot.py
from __future__ import annotations

import json
from typing import Any

from .common import run_command, text_response, ToolResponse, require_ident


def troubleshoot_deployment(args: dict[str, Any]) -> ToolResponse:
    release_name = args.get("releaseName")
    check_type = args.get("checkType", "all")
    pod_name = args.get("podName")

    try:
        namespace = require_ident(args.get("namespace"), "namespace")
        if release_name:
            release_name = require_ident(release_name, "releaseName")
        if pod_name:
            pod_name = require_ident(pod_name, "podName")

        response = f"# Troubleshooting: {namespace}\n\n"

        if check_type in ("all", "pods"):
            response += _check_pods(namespace, release_name)

        if check_type in ("all", "logs"):
            if pod_name:
                response += _get_pod_logs(namespace, pod_name)
            elif release_name:
                response += _get_pod_logs(namespace, f"{release_name}-0")

        if check_type in ("all", "helm") and release_name:
            response += _check_helm_release(namespace, release_name)

        if check_type in ("all", "route") and release_name:
            response += _check_route(namespace, release_name)

        if check_type == "vault-status" and release_name:
            response += _check_vault_status(namespace, release_name)

        response += "\n## Remediation Suggestions\n\n"
        response += _generate_remediation_suggestions(namespace)

        return text_response(response)
    except Exception as error:  # noqa: BLE001
        return text_response(f"Troubleshooting failed: {error}", is_error=True)


def _check_pods(namespace: str, release_name: str | None) -> str:
    command = ["oc", "get", "pods", "-n", namespace]
    if release_name:
        command.extend(["-l", f"app.kubernetes.io/instance={release_name}"])
    command.append("-o")
    command.append("wide")

    try:
        stdout, _ = run_command(command)
        return f"## Pod Status\n\n```\n{stdout}```\n\n"
    except Exception:  # noqa: BLE001
        return "## Pod Status\n\nUnable to retrieve pod status\n\n"


def _get_pod_logs(namespace: str, pod_name: str) -> str:
    try:
        stdout, _ = run_command(["oc", "logs", "-n", namespace, pod_name, "--tail=100"])
        return f"## Pod Logs: {pod_name}\n\n```\n{stdout}```\n\n"
    except Exception:  # noqa: BLE001
        return f"## Pod Logs: {pod_name}\n\nUnable to retrieve logs\n\n"


def _check_helm_release(namespace: str, release_name: str) -> str:
    try:
        stdout, _ = run_command(["helm", "status", release_name, "-n", namespace])
        return f"## Helm Release Status\n\n```\n{stdout}```\n\n"
    except Exception:  # noqa: BLE001
        return "## Helm Release Status\n\nRelease not found or error occurred\n\n"


def _check_route(namespace: str, release_name: str) -> str:
    try:
        stdout, _ = run_command(["oc", "get", "route", "-n", namespace, f"{release_name}-route", "-o", "wide"])
        return f"## Route Status\n\n```\n{stdout}```\n\n"
    except Exception:  # noqa: BLE001
        return "## Route Status\n\nRoute not found\n\n"


def _check_vault_status(namespace: str, release_name: str) -> str:
    pod_name = f"{release_name}-0"
    try:
        stdout, _ = run_command(["oc", "exec", "-n", namespace, pod_name, "--", "vault", "status"])
        return f"## Vault Status\n\n```\n{stdout}```\n\n"
    except Exception:  # noqa: BLE001
        return "## Vault Status\n\nUnable to check vault status (pod may not be ready)\n\n"


def _generate_remediation_suggestions(namespace: str) -> str:
    suggestions: list[str] = []
    try:
        stdout, _ = run_command(["oc", "get", "pods", "-n", namespace, "-o", "json"])
        pods = json.loads(stdout)
        items = pods.get("items", [])

        if not items:
            suggestions.append("- No pods found in namespace. Check if deployment was successful.")
        else:
            for pod in items:
                phase = pod.get("status", {}).get("phase")
                name = pod.get("metadata", {}).get("name", "unknown")
                if phase != "Running":
                    suggestions.append(f"- Pod {name} is in {phase} state")
                    for container in pod.get("status", {}).get("containerStatuses", []):
                        waiting = container.get("state", {}).get("waiting")
                        if waiting:
                            suggestions.append(f"- Container waiting: {waiting.get('reason', 'Unknown')}")
    except Exception:  # noqa: BLE001
        suggestions.append("- Unable to analyze pod status automatically")

    if not suggestions:
        suggestions.append("- All checks passed. If issues persist, check application logs.")

    return "\n".join(suggestions) + "\n"
